examples/pet-mad/pet-mad.py

```diff
@@ -0,0 +1,529 @@
+rot_angles = np.array(rot_angles)
+erot_rms = 1e3 * np.sqrt(
+    np.sum(rot_energies**2 * rot_weights) / np.sum(rot_weights)
+    - (np.sum(rot_energies * rot_weights) / np.sum(rot_weights)) ** 2
+)
+erot_max = 1e3 * np.abs(rot_energies.max() - rot_energies.min())
+print(
+    f"""
+Symmetry breaking, energy:
+RMS: {erot_rms:.3f} meV/at.
+Max: {erot_max:.3f} meV/at.
+"""
+)
+
+# %%
+# You can also inspect the rotational behavior visually
+
+chemiscope.show(
+    rot_structures,
+    mode="default",
+    properties={
+        "delta_energy": 1e3 * (rot_energies - rot_energies.mean()),
+        "euler_angles": rot_angles,
+    },
+    shapes={
+        "forces": chemiscope.ase_vectors_to_arrows(rot_structures, "forces", scale=4.0),
+    },
+    settings=chemiscope.quick_settings(
+        x="euler_angles[1]",
+        y="euler_angles[2]",
+        z="euler_angles[3]",
+        color="delta_energy",
+        structure_settings={"unitCell": True, "shape": ["forces"]},
+    ),
+)
+
+# %%
+# Note also that `i-PI <http://ipi-code.org>`_ provides functionalities to do
+# this automatically to obtain MD trajectories with a even higher
+# degree of symmetry-compliance.
+
+
+# %%
+# Simulating a complex surface
+# ----------------------------
+#
+# PET-MAD is designed to be robust and stable when executing sophisticated
+# modeling workflows. As an example, we consider a slab of an Al-6xxx alloy
+# (aluminum with a few percent Mg and Si) with some oxygen molecules adsorbed
+# at the (111) surface.
+
+al_surface = ase.io.read("data/al6xxx-o2.xyz", "0")
+
+# %%
+#
+# To run the model with external codes, it can/should be saved
+# to disk. The libraries needed to run it are collected in
+# a folder (called by default ``extensions``).
+
+model.save("pet-mad-latest.pt", collect_extensions="extensions")
+
+# %%
+# We use the ``collect_extensions`` argument to save the compiled extensions to disk.
+# These extensions ensure that the model remains self-contained and can be executed
+# without requiring the original Python or C++ source code. In particular,
+# this is necessary for the LAMMPS interface to work because it has no access to
+# the Python code.
+
+# %%
+#
+# Geometry optimization with ``ASE``
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+# As a first example, we use the ``ase`` geometry
+# `LBFGS` optimizer to relax the initial positions.
+# This leads to the rapid decomposition of the oxygen molecules
+# and the formation of an oxide layer.
+
+
+class LoggedLBFGS(LBFGS):
+    """A wrapper to LBFGS that allows saving
+    the sequence of structures along the optimization"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._traj_atoms = []
+        self._traj_energy = []
+
+    def step(self, *args, **kwargs):
+        super().step(*args, **kwargs)  # Call original step function
+
+        self._traj_atoms.append(self.atoms.copy())
+        self._traj_energy.append(self.atoms.get_potential_energy())
```

Comment:
This is not the "ASE way" to do this, IMO. I would `attach()` functions to the LBFGS object

Reply:
Feel free to change this to whatever you feel is ASE-esque enough.

Reply:
Done